src/univariate/discrete/betabinomial.jl

```diff
@@ -82,18 +82,22 @@ end
 
 function pdf(d::BetaBinomial, k::Int)
     n, α, β = d.n, d.α, d.β
-    choose = 1 / ((n+1) * beta(k+1, n-k+1))
+    chooseinv = (n+1) * beta(k+1, n-k+1)
     numerator = beta(k + α, n - k + β)
     denominator = beta(α, β)
-    return choose * (numerator / denominator)
+    return (numerator / denominator) / chooseinv
 end
 
 function pdf(d::BetaBinomial)
     n, α, β = d.n, d.α, d.β
-    k = 0:n
-    binoms = [1 / ((n+1) * beta(i+1, n-i+1)) for i in k]
-    fixed_beta = beta(α, β)
-    return binoms .* @compat(beta.(k + α, n - k + β)) / fixed_beta
+    denominator = beta(α, β)
+    values = Array{Float64, 1}()
+    for k = 0:n
+        chooseinv = (n+1) * beta(k+1, n-k+1)
+        numerator = beta(k + α, n - k + β)
+        push!(values, (numerator / denominator) / chooseinv)
```

Comment:
`push!` doesn't appear to have convert functionality, could you explain the conversion concern?
```
julia> a = Array(Float64, 0)
0-element Array{Float64,1}

julia> push!(a, "A")
ERROR: MethodError: Cannot `convert` an object of type String to an object of type Float64
This may have arisen from a call to the constructor Float64(...),
since type constructors fall back to convert methods.
 in push!(::Array{Float64,1}, ::String) at ./array.jl:479
```
I'll try to convince you otherwise about writing this like the current master syntax. I find the current syntax hard to follow. The first aspect is the nontrivial (but admittedly awesome) auto-expansion for `beta.(k + α, n - k + β)`. This broadcast of `beta` auto-expands the `k=0:n`, but I had no idea what the behavior was until I ran it to check. But if that works, why isn't `binoms = Float64[binomial(n, i) for i in k]` also written as an auto-expansion `Float64.(binomial.(n,k))`, and is the `Float64` conversion really necessary? Benchmarking the time it takes for `Float64 * Int64` and `Float64 * Float64` doesn't show any noticeable differences. I also didn't like macro in the middle of the equation. At that point, I felt I was asking too many questions and thought a simpler, more explicit method that used the same variable names as the `pdf(d::BetaBinomial, k::Int)` definition would be preferable.

Benchmarks also slightly prefer the explicit version (updated to pre-allocate the array)
```julia
using BenchmarkTools
using Distributions
using Compat
using Base.Test
import Distributions.pdf

function pdf_master(d::BetaBinomial)
	n, α, β = d.n, d.α, d.β
	k = 0:n
	binoms = Float64[binomial(n, i) for i in k]
	fixed_beta = beta(α, β)
	return binoms .* @compat(beta.(k + α, n - k + β)) / fixed_beta
end

function pdf_elementwise(d::BetaBinomial)
    n, α, β = d.n, d.α, d.β
	k = 0:n
	binoms = [(n+1) * beta(i+1, n-i+1) for i in k]
	fixed_beta = beta(α, β)
	return @compat(beta.(k + α, n - k + β)) ./ (fixed_beta .* binoms)
end

function pdf_preinitialized(d::BetaBinomial)
    n, α, β = d.n, d.α, d.β
    denominator = beta(α, β)
    values = Array(Float64, n+1)
    for (i,k) in enumerate(0:n)
        chooseinv = (n+1) * beta(k+1, n-k+1)
        numerator = beta(k + α, n - k + β)
        values[i] = numerator / (denominator * chooseinv)
    end
    return values
end
```
```julia
julia> # check accuracy
       d = BetaBinomial(3, 2, 2)
Distributions.BetaBinomial{Float64}(n=3, α=2.0, β=2.0)

julia> answers = [0.19999999999999998, 0.29999999999999977, 0.29999999999999977, 0.19999999999999998]
4-element Array{Float64,1}:
 0.2
 0.3
 0.3
 0.2

julia> @test isapprox(pdf_master(d), answers)
Test Passed
  Expression: isapprox(pdf_master(d),answers)

julia> @test isapprox(pdf_elementwise(d), answers)
Test Passed
  Expression: isapprox(pdf_elementwise(d),answers)

julia> @test isapprox(pdf_preinitialized(d), answers)
Test Passed
  Expression: isapprox(pdf_preinitialized(d),answers)

julia> # check efficiency
       d = BetaBinomial(10,2,2)
Distributions.BetaBinomial{Float64}(n=10, α=2.0, β=2.0)

julia> @benchmark pdf_master(d)
BenchmarkTools.Trial:
  samples:          10000
  evals/sample:     9
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  4.14 kb
  allocs estimate:  44
  minimum time:     3.00 μs (0.00% GC)
  median time:      3.06 μs (0.00% GC)
  mean time:        3.46 μs (7.92% GC)
  maximum time:     233.08 μs (96.08% GC)

julia> @benchmark pdf_elementwise(d)
BenchmarkTools.Trial:
  samples:          10000
  evals/sample:     7
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  7.23 kb
  allocs estimate:  77
  minimum time:     4.32 μs (0.00% GC)
  median time:      4.47 μs (0.00% GC)
  mean time:        5.17 μs (8.21% GC)
  maximum time:     312.74 μs (94.84% GC)

julia> @benchmark pdf_preinitialized(d)
BenchmarkTools.Trial:
  samples:          10000
  evals/sample:     7
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  6.64 kb
  allocs estimate:  70
  minimum time:     4.07 μs (0.00% GC)
  median time:      4.20 μs (0.00% GC)
  mean time:        4.70 μs (6.78% GC)
  maximum time:     308.45 μs (96.60% GC)

julia> # check larger scale efficiency
       d = BetaBinomial(1029,2,2)
Distributions.BetaBinomial{Float64}(n=1029, α=2.0, β=2.0)

julia> @benchmark pdf_elementwise(d)
BenchmarkTools.Trial:
  samples:          9779
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  612.48 kb
  allocs estimate:  6191
  minimum time:     440.05 μs (0.00% GC)
  median time:      456.44 μs (0.00% GC)
  mean time:        510.72 μs (7.88% GC)
  maximum time:     2.74 ms (71.10% GC)

julia> @benchmark pdf_preinitialized(d)
BenchmarkTools.Trial:
  samples:          10000
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  587.84 kb
  allocs estimate:  6184
  minimum time:     394.74 μs (0.00% GC)
  median time:      423.55 μs (0.00% GC)
  mean time:        471.70 μs (8.04% GC)
  maximum time:     2.84 ms (68.21% GC)
```